flows/knack/knack_services.py

```diff
@@ -0,0 +1,400 @@
+    # Let's configure the agents to download the file from this repo
+    storage=GitHub(
+        repo="cityofaustin/atd-prefect",
+        path="flows/knack/knack_services.py",
+        ref="atd-knack-services",  # The branch name
+    ),
+    run_config=LocalRun(labels=["atd-data02", "production"]),
+    schedule=None,
+    state_handlers=[set_run_name],
+) as flow:
+    # Parameter tasks
+    app_name = Parameter("App Name", default=APP_NAME, required=True)
+    container = Parameter("Knack Container", default=CONTAINER, required=True)
+    layer = Parameter(
+        "AGOL Build Segment Geometry Layer", default=LAYER_NAME, required=False
+    )
+    replace_data = Parameter("Replace all Data", default=REPLACE_DATA, required=True)
+    app_name_dest = Parameter(
+        "Records to Knack: App Name Destination", default=APP_NAME_DEST, required=False
+    )
+    soda_flag = Parameter("Send data to Socrata", default=True, required=True)
+    agol_flag = Parameter("Send data to AGOL", default=True, required=True)
+
+    # Based on provided parameters, skip or run some conditional tasks
+    build_geom, to_knack = determine_task_runs(layer, app_name_dest)
+
+    # Get the last time the flow ran for this app/container combo
+    date_filter = get_last_exec_time(app_name, container, replace_data)
+
+    environment_variables = get_env_vars(app_name)
+
+    # 1. Pull latest docker image
+    docker_pull = pull_docker_image()
+    # 2. Download Knack records and send them to Postgres(t)
+    postgrest_res = records_to_postgrest(
+        app_name,
+        container,
+        date_filter,
+        environment_variables,
+        upstream_tasks=[docker_pull],
+    )
+    # 3. Send data from Postgrest to AGOL (optional)
+    with case(agol_flag, True):
+        agol_res = records_to_agol(
+            app_name,
+            container,
+            date_filter,
+            environment_variables,
+            upstream_tasks=[docker_pull, postgrest_res],
+        )
+
+    # 4. Send data from Postgrest to Socrata (optional)
+    with case(soda_flag, True):
+        socrata_res = records_to_socrata(
+            app_name,
+            container,
+            date_filter,
+            environment_variables,
+            upstream_tasks=[docker_pull, postgrest_res],
+        )
+    # 5. Build line geometries in AGOL (optional)
+    with case(build_geom, True):
+        agol_build_res = agol_build_markings_segment_geometries(
+            layer,
+            date_filter,
+            environment_variables,
+            upstream_tasks=[docker_pull, agol_res, build_geom],
+        )
+    # 6. Send data to another knack app (optional)
+    with case(to_knack, True):
+        knack_res = records_to_knack(
+            app_name,
+            container,
+            date_filter,
+            app_name_dest,
+            environment_variables,
+            upstream_tasks=[docker_pull, postgrest_res, to_knack],
+        )
+
```

Comment:
I wonder if it would be helpful to `merge()` all of these `case()`es results back into a single node on the DAG and make it an upstream node of `update_last_exec_time()`?

Reply:
I haven't use `merge()` before, also sorry for responding super late on this comment @frankhereford. I think merge all of these `upstream_tasks` into one task? I'm not even sure I need `upstream_tasks` here anyway since they are now supplied to our task arguments now.